cpp/src/arrow/engine/substrait/util.cc

```diff
@@ -68,7 +68,7 @@ class SubstraitExecutor {
                              compute::ExecContext exec_context)
       : plan_(std::move(plan)), exec_context_(exec_context) {}
 
-  ~SubstraitExecutor() { ARROW_CHECK_OK(this->Close()); }
+  ~SubstraitExecutor() { ARROW_UNUSED(this->Close()); }
```

Comment:
It gets checked [earlier here](https://github.com/apache/arrow/pull/13375/files#diff-0078df867c0f64dcfe79c149ee007f0933d164a32f122e28d79872a969b5be05R121). I moved the check there because I observed that the original code crashed (in this destructor) when there was some error during the execution of the plan. Such errors happen normally, and I wanted explanations for errors to surface.

Regarding a warning in the destructor, I could add it if you still want it. Note that an error would get logged even if the user handled the error in the [earlier call to `ExecuteSerializedPlan`](https://github.com/apache/arrow/pull/13375/files#diff-0078df867c0f64dcfe79c149ee007f0933d164a32f122e28d79872a969b5be05R111), and I'm not sure this is what users would like.

Reply:
In that case, `Close()` should probably be fixed so that it's idempotent/we should have some flag to tell that we don't need to close

Reply:
I suppose idempotent is not the right word.

I guess I can see the motivation for this behavior: we can just swallow the error, and it won't matter to "well behaved" application since it will have handled the error, and then we don't need to track any extra state. So I suppose this is ok.

Reply:
I agree. Also note that the flag-solution could still lead to a crash in the destructor if `Close()` fails there and was not called earlier. Besides avoiding a crash, the proposed `SubstraitExecutor` code ensures that `ExecuteSerializedPlan` calls `Close()` or reports some error back or both.